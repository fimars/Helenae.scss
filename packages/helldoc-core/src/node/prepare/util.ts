import fs = require("fs-extra");
import * as path from "path";

const tempCache = new Map();

export function createTemp(tempPath?: string) {
  if (!tempPath) {
    tempPath = path.resolve(__dirname, "../../.temp");
  } else {
    tempPath = path.resolve(tempPath);
  }

  if (!fs.existsSync(tempPath)) {
    fs.ensureDirSync(tempPath);
  } else {
    fs.emptyDirSync(tempPath);
  }

  async function writeTemp(file: string, content: string | Buffer) {
    const destPath = path.join(tempPath as string, file);
    await fs.ensureDir(path.parse(destPath).dir);

    const cached = tempCache.get(file);
    if (cached !== content) {
      await fs.writeFile(destPath, content);
      tempCache.set(file, content);
    }
    return destPath;
  }

  return { writeTemp, tempPath };
}

export function isIndexFile(file: string) {
  return /^(index|readme)\.md$/i.test(file);
}
