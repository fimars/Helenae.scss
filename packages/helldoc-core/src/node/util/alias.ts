import { resolve } from "path";

export function resolveApp(path: string) {
  return resolve(__dirname, "../../app", path);
}
