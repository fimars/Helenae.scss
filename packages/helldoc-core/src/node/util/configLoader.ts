import { existsSync } from "fs-extra";
import { resolve } from "path";

// TODO: use egoist/joycon support multiple configuration types.
export function loadConfig(name: string, cwd: string) {
  const configPath = resolve(cwd, name);
  if (existsSync(configPath)) {
    return require(configPath) || {};
  }
  return {};
}
