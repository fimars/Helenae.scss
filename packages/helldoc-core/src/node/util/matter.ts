import * as toml from "toml";
import * as matter from "gray-matter";

export function parseFrontmatter(content: string) {
  return matter(content.trim(), {
    engines: {
      toml: toml.parse.bind(toml)
    },
    excerpt: true,
    excerpt_separator: "<!-- more -->"
  });
}
