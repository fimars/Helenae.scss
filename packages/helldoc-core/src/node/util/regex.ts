export function isJS(str: string) {
  return /.+\.js$/.test(str);
}
