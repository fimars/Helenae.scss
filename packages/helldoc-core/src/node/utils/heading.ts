import * as marked from "marked";
import { TocHeader } from "../../types";

interface Header extends marked.Tokens.Heading {
  id: string;
  parent: string | null;
}

export function extractHeaders(
  content: string,
  parser: any = marked.lexer
): TocHeader[] {
  const tokens: marked.TokensList = parser(content);
  const headingTokens = tokens.filter(
    token => token.type === "heading"
  ) as Header[];
  const slugger = new marked.Slugger();

  const headers: TocHeader[] = headingTokens.map(({ depth, text }) => ({
    id: slugger.slug(text),
    level: depth,
    text,
    parent: null
  }));
  headers.forEach((header, idx) => {
    header.parent = findHeaderParent(header.level, idx - 1, headers);
  });
  return headers;
}

function findHeaderParent(
  level: number,
  idx: number,
  tokens: TocHeader[]
): string | null {
  for (; idx >= 0; idx--) {
    const prevToken = tokens[idx];
    if (level > prevToken.level) {
      return prevToken.id;
    }
    if (prevToken.level === level) {
      return prevToken.parent || null;
    }
  }
  return null;
}
