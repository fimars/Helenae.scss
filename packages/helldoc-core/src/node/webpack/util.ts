import resolveFrom = require("resolve-from");
import { resolve, dirname } from "path";
import { existsSync } from "fs-extra";

export function resolvePackage(
  name: string,
  { cwd = process.cwd(), prefix = "" }
) {
  const packageDir = dirname(resolveFrom(cwd, `${prefix}${name}/package.json`));
  const buildDir = `${packageDir}/build`;
  const distDir = `${packageDir}/build`;
  if (existsSync(distDir)) return distDir;
  if (existsSync(buildDir)) return buildDir;
  return packageDir;
}

export function getModulePaths(cwd = process.cwd()) {
  return [resolve(cwd, "node_modules")].concat(module.paths);
}
