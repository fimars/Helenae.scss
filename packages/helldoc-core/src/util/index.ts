import * as matter from "gray-matter";
import { TokensList, Tokens, Slugger } from "marked";
import { TocHead } from "../types";
import toml = require("toml");

interface HellHeader extends Tokens.Heading {
  id?: string;
  parent?: string;
}
export function extractHeaders(content: string, lexer: any): TocHead[] {
  const tokens = lexer(content) as TokensList;
  const headerTokens = tokens.filter(
    ({ type }) => type === "heading"
  ) as HellHeader[];
  const slugger = new Slugger();
  const headers = headerTokens.map(({ depth, text }, headerIndex) => {
    const parent = findHeaderParent(depth, headerIndex, headerTokens);
    return {
      id: slugger.slug(text),
      level: depth,
      parent,
      text
    };
  });
  return headers;
}
function findHeaderParent(
  depth: number,
  idx: number,
  headers: HellHeader[]
): string | null {
  for (; idx > 0; idx--) {
    const { depth: prevDepth, id: prevId, parent: prevParent } = headers[idx];
    if (depth > prevDepth) {
      return prevId as string;
    }
    if (prevParent && prevDepth === depth) {
      return prevParent;
    }
  }
  return null;
}

export function parseFrontmatter(content: string) {
  return matter(content.trim(), {
    engines: {
      toml: toml.parse.bind(toml)
    },
    excerpt: true,
    excerpt_separator: "<!-- more -->"
  });
}
