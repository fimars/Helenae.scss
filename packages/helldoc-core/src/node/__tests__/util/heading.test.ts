import { extractHeaders } from "../../util/heading";

it("extract headers correctly", () => {
  const headers = extractHeaders(
    `
# t1
## t1-1
## t1-2
### t1-2-1
# t2
## t2-1
### t2-1-1
## t2-2
  `.trim()
  );
  expect(headers).toMatchSnapshot();
});
