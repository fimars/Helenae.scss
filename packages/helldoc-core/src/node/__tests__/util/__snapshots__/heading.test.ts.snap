// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract headers correctly 1`] = `
Array [
  Object {
    "id": "t1",
    "level": 1,
    "parent": null,
    "text": "t1",
  },
  Object {
    "id": "t1-1",
    "level": 2,
    "parent": "t1",
    "text": "t1-1",
  },
  Object {
    "id": "t1-2",
    "level": 2,
    "parent": "t1",
    "text": "t1-2",
  },
  Object {
    "id": "t1-2-1",
    "level": 3,
    "parent": "t1-2",
    "text": "t1-2-1",
  },
  Object {
    "id": "t2",
    "level": 1,
    "parent": null,
    "text": "t2",
  },
  Object {
    "id": "t2-1",
    "level": 2,
    "parent": "t2",
    "text": "t2-1",
  },
  Object {
    "id": "t2-1-1",
    "level": 3,
    "parent": "t2-1",
    "text": "t2-1-1",
  },
  Object {
    "id": "t2-2",
    "level": 2,
    "parent": "t2",
    "text": "t2-2",
  },
]
`;
